File: 0733-flood-fill/0733-flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        target = image[sr][sc]
        
        def flood_fill(x, y):
            if x < 0 or x >= len(image): return
            if y < 0 or y >= len(image[0]): return

            if image[x][y] == color: return
            if image[x][y] != target: return

            image[x][y] = color

            flood_fill(x-1, y)
            flood_fill(x+1, y)
            flood_fill(x, y+1)
            flood_fill(x, y-1)

        flood_fill(sr, sc)
        return image

File: 0733-flood-fill/0733-flood-fill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        if (image[sr][sc] == color) return image;
        flood(image, sr, sc, image[sr][sc], color);
        return image;
    }

    public void flood(int[][] image, int sr, int sc, int color, int newColor) {
        if (sr < 0 || sc < 0 || sr >= image.length || sc >= image[0].length || image[sr][sc] != color) return;
        image[sr][sc] = newColor;

        flood(image, sr - 1, sc, color, newColor);
        flood(image, sr, sc - 1, color, newColor);
        flood(image, sr + 1, sc, color, newColor);
        flood(image, sr, sc + 1, color, newColor);
    }
}